.github/actions/detect-workflow-js/src/detect.ts

```diff
@@ -134,3 +135,58 @@ export async function detectWorkflowFromContext(
 
   return [repository, ref, workflow];
 }
+
+/**
+ Confirm that all jobs in the calling workflow are not using self-hosted Runners.
+ This check is only needed when calling from the generic builders, check if we're using a "generic builder", 
+ where the user may attempt to supply artifacts produced on self-hosted runners.
+
+ This check requires the caller to specify a github token with an additional `administration:read` permissions.
+ example:
+ ```
+ uses: slsa-framewrok/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml'
+   secrets:
+     token: ${{ secrets.MY_TOKEN_WITH_EXTRA_PERM }}
+  ```
+*/
+export async function ensureOnlyGithubHostedRunners(
+  repoName: string,
+  token: string,
+): Promise<void> {
+  const [owner, repo] = repoName.split("/");
```

Comment:
do you know what permissions are needed to read this API? Does it require admin permissions?

Reply:
Yes, it needs a user-generated token with actions: read and administration:read against the repo. In the new action I'll write up the docs for it.

Reply:
I'm worried about this requirement `admin:read`. This is a hard ask. @ianlewis @kpk47 @pnacht @haydentherapper wdut?

Reply:
This would give me pause if I saw I needed to generate an admin PAT. Which of these APIs needs this permission?

Reply:
`admin: read` and `actions: read`, but the user can limit their token's permissions to a single repo.

Reply:
> Which of these APIs needs this permission?

`octokit.rest.actions.listSelfHostedRunnersForRepo` at least needs it.

But I agree, requiring this permission can be a hard sell...

---

Question: can't this permission be replaced by having a list of [known GitHub runner images](https://github.com/actions/runner-images?tab=readme-ov-file#available-images)?

Then, if `runs-on` contains only elements in that list, it's assumed to be a GitHub image.

I don't know if a self-hosted image with label `ubuntu-latest` will ever be chosen... But even if so, I assume it's highly unreliable (you never know if your runner will be chosen before the GH runner), unless GH gives preference to self-hosted runners whenever possible.

This can either be hardcoded into the generator (therefore requiring updates whenever GitHub releases a new image) or be dynamically parsed from the README (fragile!). I unfortunately couldn't find a REST API endpoint or something of the sort we can query to get all the images.

Reply:
That's right. Github doesn't publish a full list, and if I assign my runner with the label "ubuntu-latest" or "macos-14-large", then my job can still run on my self-hosted runner. There's nothing stopping me from assigning my runner with the labels that github already uses.

Reply:
@steiza Any guidance how we can figure out whether a run uses self-hosted runner vs GitHub-hosted runners, and without admin permissions - https://github.com/slsa-framework/slsa-github-generator/issues/1868?

Reply:
@josepalafox @patrick-knight

Reply:
I've also opened a ticket with github

- https://github.com/orgs/community/discussions/111347
- https://github.com/slsa-framework/slsa-github-generator/issues/1868#issuecomment-1985901738